File: frontend/js/appointments.js
/* eSHCAT — appointments.js
   Appointment requests. */

(function () {
  const ESH = window.ESH;

  async function seedOfficeSelect(selectId, serviceSelectId) {
    const select = document.getElementById(selectId);
    const serviceSelect = document.getElementById(serviceSelectId);
    if (!select) return;
    let services = [];
    try {
      services = await ESH.api.get("/services");
      const departments = {};
      services.forEach((s) => (departments[s.department_id] = s.department));
      Object.entries(departments).forEach(([id, name]) => {
        const option = ESH.el("option", { value: id, text: name });
        select.appendChild(option);
      });
    } catch {
      /* leave empty; backend validates */
    }

    if (serviceSelect) {
      select.addEventListener("change", () => {
        serviceSelect.innerHTML = "";
        serviceSelect.appendChild(ESH.el("option", { value: "", text: "Any service for this office..." }));
        const deptId = select.value;
        services
          .filter((s) => String(s.department_id) === String(deptId))
          .forEach((s) => {
            serviceSelect.appendChild(ESH.el("option", { value: s.id, text: s.name }));
          });
      });
    }
  }

  async function init() {
    const form = document.getElementById("appointmentForm");
    if (!form) return;
    await seedOfficeSelect("deptSelect", "apptService");

    form.addEventListener("submit", async (e) => {
      e.preventDefault();
      const btn = document.getElementById("apptBtn");
      const payload = {
        department_id: document.getElementById("deptSelect").value || null,
        service_id: document.getElementById("apptService").value || null,
        date: document.getElementById("apptDate").value,
        time: document.getElementById("apptTime").value,
        full_name: document.getElementById("apptName").value,
        email: document.getElementById("apptEmail").value,
        mobile: document.getElementById("apptMobile").value,
      };

      if (!payload.department_id) {
        ESH.showToast("Please select an office.", "error");
        return;
      }

      ESH.setLoading(btn, "Requesting...");
      try {
        const data = await ESH.api.post("/appointments", payload);
        const formContainer = form.parentElement;
        form.remove();
        const card = ESH.el("div", { class: "card", style: "max-width:560px;margin:0 auto;text-align:center" }, [
          ESH.el("div", { style: "font-size:2.4rem;color:var(--color-success)", text: "" }, [ESH.el("i", { class: "fa-solid fa-circle-check", "aria-hidden": "true" })]),
          ESH.el("h2", { text: "Appointment Requested" }),
          ESH.el("p", { class: "note", text: "Keep your reference number for your visit." }),
          ESH.el("div", { class: "ref-box" }, [
            ESH.el("div", { class: "ref", text: data.reference_number }),
          ]),
          ESH.el("div", { style: "display:flex;gap:10px;justify-content:center;flex-wrap:wrap" }, [
            ESH.el("a", { class: "btn btn-primary", href: "/", text: "Back to Home" }),
            ESH.el("a", { class: "btn btn-secondary", href: "/pages/track.html", text: "Track a Request" }),
          ]),
        ]);
        formContainer.appendChild(card);
        ESH.showToast("Appointment requested.", "success");
      } catch (err) {
        ESH.unsetLoading(btn);
        ESH.showToast(err.message, "error");
      }
    });
  }

  document.addEventListener("DOMContentLoaded", () => {
    if (document.body.dataset.page === "appointments") init();
  });
})();

File: backend/routes/appointments.py
from flask import Blueprint, jsonify, request

from ..database import get_connection
from ..services.reference_service import generate_reference

appointments_bp = Blueprint("appointments", __name__)


@appointments_bp.route("/api/appointments", methods=["POST"])
def request_appointment():
    payload = request.get_json(silent=True) or {}
    department_id = payload.get("department_id")
    service_id = payload.get("service_id")
    date = payload.get("date")
    time = payload.get("time")
    full_name = (payload.get("full_name") or "").strip()
    email = (payload.get("email") or "").strip()
    mobile = payload.get("mobile") or ""

    if not full_name or not email or not date or not time:
        return jsonify({"error": "Name, email, date and time are required."}), 400

    conn = get_connection()
    try:
        reference = generate_reference("CAT-APT")
        conn.execute(
            """
            INSERT INTO appointments
                (reference_number, department_id, service_id, appointment_date, appointment_time, full_name, email, mobile)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (reference, department_id, service_id, date, time, full_name, email, mobile),
        )
        conn.commit()
        return jsonify(
            {
                "message": "Appointment requested.",
                "reference_number": reference,
            }
        ), 201
    finally:
        conn.close()